File: src/data/human/Find_the_Median_of_the_Uniqueness_Array/Find_the_Median_of_the_Uniqueness_Array.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int medianOfUniquenessArray(vector<int>& A) {
        int n = A.size();
        long long total = (long long) n * (n + 1) / 2;

        auto atmost = [&](int k) {
            long long res = 0;
            unordered_map<int, int> count;
            for (int i = 0, j = 0; j < n; j++) {
                count[A[j]]++;
                while (count.size() > k) {
                    if (--count[A[i]] == 0)
                        count.erase(A[i]);
                    i++;
                }
                res += j - i + 1;
            }
            return res;
        };

        int left = 1, right = unordered_set<int>(A.begin(), A.end()).size();
        while (left < right) {
            int k = (left + right) / 2;
            if (atmost(k) * 2 >= total)
                right = k;
            else
                left = k + 1;
        }
        return left;
    }
};


//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Small array with unique elements
    std::vector<int> nums = {1, 2, 3};
    int expected_result = 1;  // Subarrays have distinct counts [1, 1, 1, 2, 2, 3], median is 1
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 1 failed.");

    // Test case 2: Array with repeated elements
    nums = {1, 1, 2};
    expected_result = 1;  // Subarrays have distinct counts [1, 1, 2, 1, 2, 2], median is 1
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 2 failed.");

    // Test case 3: All identical elements
    nums = {4, 4, 4, 4};
    expected_result = 1;  // All subarrays have 1 unique element
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 3 failed.");

    // Test case 4: Array with two distinct elements
    nums = {5, 6, 5, 6};
    expected_result = 2;  // Subarrays have distinct counts [1, 1, 2, 1, 2, 2, 2, 2], median is 2
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 4 failed.");

    // Test case 5: Increasing sequence of elements
    nums = {1, 2, 3, 4};
    expected_result = 2;  // Subarrays have distinct counts [1, 1, 1, 1, 2, 2, 2, 3, 3, 4], median is 2
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 5 failed.");

    // Test case 6: Mixed values with multiple subarrays having the same uniqueness count
    nums = {2, 1, 2, 3};
    expected_result = 2;  // Subarrays have distinct counts [1, 1, 2, 1, 2, 2, 2, 3, 3, 4], median is 2
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 6 failed.");

    // Test case 7: Array with no repeats
    nums = {3, 4, 3, 4, 5};
    expected_result = 2;  // Subarrays have distinct counts [1, 1, 1, 2, 2, 3], median is 2
    assert(solution.medianOfUniquenessArray(nums) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 327680; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/human/Find_the_Median_of_the_Uniqueness_Array/Find_the_Median_of_the_Uniqueness_Array.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int medianOfUniquenessArray(int[] A) {
        int n = A.length, left = 1, right = n;
        long total = (long) n * (n + 1) / 2;
        while (left < right) {
            int k = (left + right) / 2;
            if (atmost(A, k) * 2 >= total) {
                right = k;
            } else {
                left = k + 1;
            }
        }
        return left;
    }

    private long atmost(int[] A, int k) {
        long res = 0;
        Map<Integer, Integer> count = new HashMap<>();
        int i = 0;
        for (int j = 0; j < A.length; j++) {
            count.put(A[j], count.getOrDefault(A[j], 0) + 1);
            while (count.size() > k) {
                count.put(A[i], count.get(A[i]) - 1);
                if (count.get(A[i]) == 0) {
                    count.remove(A[i]);
                }
                i++;
            }
            res += j - i + 1;
        }
        return res;
    }
}

public class Find_the_Median_of_the_Uniqueness_Array {
public static void runTests(Solution solution) {
    // Test case 1: Small array with unique elements
    int[] nums1 = {1, 2, 3};
    int expected_result1 = 1;  // Subarrays have distinct counts, median is 1
    assert solution.medianOfUniquenessArray(nums1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Array with repeated elements
    int[] nums2 = {1, 1, 2};
    int expected_result2 = 1;  // Subarrays have distinct counts, median is 1
    assert solution.medianOfUniquenessArray(nums2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: All identical elements
    int[] nums3 = {4, 4, 4, 4};
    int expected_result3 = 1;  // All subarrays have 1 unique element
    assert solution.medianOfUniquenessArray(nums3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Array with two distinct elements
    int[] nums4 = {5, 6, 5, 6};
    int expected_result4 = 2;  // Subarrays have distinct counts, median is 2
    assert solution.medianOfUniquenessArray(nums4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Increasing sequence of elements
    int[] nums5 = {1, 2, 3, 4};
    int expected_result5 = 2;  // Subarrays have distinct counts, median is 2
    assert solution.medianOfUniquenessArray(nums5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Mixed values with multiple subarrays having the same uniqueness count
    int[] nums6 = {2, 1, 2, 3};
    int expected_result6 = 2;  // Subarrays have distinct counts, median is 2
    assert solution.medianOfUniquenessArray(nums6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Array with no repeats
    int[] nums7 = {3, 4, 3, 4, 5};
    int expected_result7 = 2;  // Subarrays have distinct counts, median is 2
    assert solution.medianOfUniquenessArray(nums7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 10000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    